cmd/kubeadm/app/cmd/join_test.go

```diff
@@ -243,3 +246,233 @@ func TestNewJoinData(t *testing.T) {
 		})
 	}
 }
+
+func initJoinOptions(token string, discovery kubeadmapiv1beta1.Discovery, controlPlane bool) *joinOptions {
+	joinOptions := newJoinOptions()
+	joinOptions.token = token
+	joinOptions.externalcfg.Discovery = discovery
+	joinOptions.controlPlane = controlPlane
+	return joinOptions
+}
+
+func TestAmendExternalJoinConfiguration(t *testing.T) {
+	var tests = []struct {
+		description  string
+		args         []string
+		flags        map[string]string
+		token        string
+		controlPlane bool
+		discovery    kubeadmapiv1beta1.Discovery
+		expectedErr  bool
+		validate     func(*testing.T, *joinOptions)
+	}{
+		{
+			description: "len(opt.token) = 0",
+			args:        []string{"1.2.3.4:6443"},
+			flags:       map[string]string{options.FileDiscovery: "https://foo"},
+			token:       "",
+			discovery: kubeadmapiv1beta1.Discovery{
+				BootstrapToken: &kubeadmapiv1beta1.BootstrapTokenDiscovery{
+					Token: "abcdef.0123456789abcdef",
+				},
+				File:              &kubeadmapiv1beta1.FileDiscovery{},
+				TLSBootstrapToken: "abcdef.0123456789",
+			},
+			expectedErr: false,
+			validate: func(t *testing.T, opt *joinOptions) {
```

Comment:
I am thinking, that it may be better to just have an "expected" external cfg here and use `reflect.DeepEqual` to check that the result is the same as expected. That way we can check for the potential impact on irrelevant fields (the validate funcs only verify the fields of interest).

Reply:
good idea, the problem is the return of reflect.DeepEqual is bool, we will can't know exactly which field is wrong, and the efficiency of reflect.DeepEqual is not very well.

Reply:
DeepEqual is slow, but it works for tests. It's not mandatory to use it, but if you don't you'll have to check a few more things here.